stats/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,239 @@
+// pick a subset of metrics, use this and add onto it.
+var EmptyMetrics = Metrics{}
+
+// Metrics is a set of metrics to initialize. Once created, Metrics is
+// immutable.
+type Metrics struct {
+	// metrics are the set of metrics to initialize.
+	metrics map[string]bool
+}
+
+// Add adds the metrics to the metrics set and returns a new copy with the
+// additional metrics.
+func (m *Metrics) Add(metrics ...string) *Metrics {
+	newMetrics := make(map[string]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		newMetrics[metric] = true
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// Remove removes the metrics from the metrics set and returns a new copy with
+// the metrics removed.
+func (m *Metrics) Remove(metrics ...string) *Metrics {
+	newMetrics := make(map[string]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		delete(newMetrics, metric)
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// MetricsOptions are the metrics options for OpenTelemetry instrumentation.
+type MetricsOptions struct {
+	// MeterProvider is the MeterProvider instance that will be used for access
+	// to Named Meter instances to instrument an application. To enable metrics
+	// collection, set a meter provider. If unset, no metrics will be recorded.
+	// Any implementation knobs (i.e. views, bounds) set in the passed in object
+	// take precedence over the API calls from the interface in this component
+	// (i.e. it will create default views for unset views).
+	MeterProvider metric.MeterProvider
+	// Metrics are the metrics to instrument. Will turn on the corresponding
+	// metric supported by the client and server instrumentation components if
+	// applicable.
+	Metrics Metrics
+	// TargetAttributeFilter is a callback that takes the target string and
+	// returns a bool representing whether to use target as a label value or use
+	// the string "other". If unset, will use the target string as is.
+	TargetAttributeFilter func(string) bool
+	// MethodAttributeFilter is a callback that takes the method string and
+	// returns a bool representing whether to use method as a label value or use
+	// the string "other". If unset, will use the method string as is. This is
+	// used only for generic methods, and not registered methods.
+	MethodAttributeFilter func(string) bool
+}
+
+// DialOption returns a dial option which enables OpenTelemetry instrumentation
+// code for a grpc.ClientConn.
+//
+// Client applications interested in instrumenting their grpc.ClientConn should
+// pass the dial option returned from this function as a dial option to
+// grpc.Dial().
+//
+// For the metrics supported by this instrumentation code, a user needs to
+// specify the client metrics to record in metrics options. A user also needs to
+// provide an implementation of a MeterProvider. If the passed in Meter Provider
```

Comment:
Should this check for `MeterProvider == nil` and `panic` if so?  (I would suggest "return an error", but it's nice if this just returns a `DialOption` only.

Reply:
I only vaugely remember this, but me and Yash talked about this last summer. There was a concept of a "no-op" meter provider, and we went back and forth on the logic. There was some reason why Yash wanted a Meter Provider that does nothing. Rather then check for it, we decided that no Meter Provider = a no-op Open Telemetry instrumentation component.

Reply:
@yashykt to see if he remembers this conversation.

Reply:
Yash just mentioned one big reason is we will eventually support tracing, thus you can't assert here since you can configure tracing with no metrics.

Reply:
correct